File: frontend/src/services/api/experiment.ts
import { apiV1Get, apiV1Post, apiV1Put } from '@/services/client'
import { IResponse, IWithPagination } from '@/services/types'

export type ExperimentVisibility = 'private' | 'account'
export type ExperimentRunStatus =
  | 'pending'
  | 'running'
  | 'succeeded'
  | 'failed'
  | 'terminated'
  | 'cancelled'

export interface Experiment {
  ID: number
  CreatedAt: string
  UpdatedAt: string
  name: string
  description: string
  userID: number
  accountID: number
  visibility: ExperimentVisibility
  tags?: Record<string, unknown>
}

export interface ExperimentRun {
  ID: number
  CreatedAt: string
  UpdatedAt: string
  experimentID: number
  parentRunID?: number
  sourceCheckpointID?: number
  jobID?: number
  jobName: string
  runName: string
  status: ExperimentRunStatus
  userID: number
  accountID: number
  hyperparams?: Record<string, unknown>
  codeSnapshot?: Record<string, unknown>
  dataSnapshot?: Record<string, unknown>
  imageSnapshot?: Record<string, unknown>
  resourceSnapshot?: Record<string, unknown>
  checkpointSnapshot?: Record<string, unknown>
  reproduceSnapshot?: Record<string, unknown>
  tags?: Record<string, unknown>
  startedAt?: string
  finishedAt?: string
}

export interface RunMetric {
  ID: number
  CreatedAt: string
  runID: number
  clientRecordID?: string
  name: string
  step: number
  value: number
  timestamp: string
  context?: Record<string, unknown>
}

export interface RunMetricQuery {
  names?: string[]
  startStep?: number
  endStep?: number
  limit?: number
  downsample?: number
}

export interface RunArtifact {
  ID: number
  CreatedAt: string
  runID: number
  clientRecordID?: string
  name: string
  type: string
  path: string
  sizeBytes: number
  sourceType?: string
  sourceID?: number
  metadata?: Record<string, unknown>
}

export interface ExperimentCreateReq {
  name: string
  description?: string
  visibility?: ExperimentVisibility
  tags?: Record<string, unknown>
}

export interface ExperimentRunConfig {
  experimentId: number
  runName?: string
  hyperparams?: Record<string, unknown>
  code?: Record<string, unknown>
  data?: Record<string, unknown>
  image?: Record<string, unknown>
  tags?: Record<string, unknown>
}

export interface CheckpointRestorePlan {
  jobName: string
  checkpointID: number
  name: string
  checkpointPath: string
}

export interface ReproduceRunResult {
  mode: 'checkpoint-restore' | 'snapshot'
  runID: number
  experimentID: number
  runName: string
  restore?: CheckpointRestorePlan
  createJobExperimentPayload: ExperimentRunConfig
  snapshots: Record<string, unknown>
}

export const apiExperimentList = () =>
  apiV1Get<IResponse<IWithPagination<Experiment>>>('experiments')

export const apiExperimentCreate = (payload: ExperimentCreateReq) =>
  apiV1Post<IResponse<Experiment>>('experiments', payload)

export const apiExperimentGet = (id: number) => apiV1Get<IResponse<Experiment>>(`experiments/${id}`)

export const apiExperimentUpdate = (id: number, payload: Partial<ExperimentCreateReq>) =>
  apiV1Put<IResponse<Experiment>>(`experiments/${id}`, payload)

export const apiExperimentRuns = (id: number) =>
  apiV1Get<IResponse<ExperimentRun[]>>(`experiments/${id}/runs`)

export const apiExperimentRunGet = (runID: number) =>
  apiV1Get<IResponse<ExperimentRun>>(`experiments/runs/${runID}`)

export const apiExperimentRunMetrics = (runID: number, query?: RunMetricQuery) => {
  const searchParams = compactMetricQuery(query)
  return apiV1Get<IResponse<RunMetric[]>>(
    `experiments/runs/${runID}/metrics`,
    searchParams ? { searchParams } : undefined
  )
}

export const apiExperimentRunArtifacts = (runID: number) =>
  apiV1Get<IResponse<RunArtifact[]>>(`experiments/runs/${runID}/artifacts`)

export const apiExperimentRunReproduce = (runID: number, payload?: { name?: string }) =>
  apiV1Post<IResponse<ReproduceRunResult>>(`experiments/runs/${runID}/reproduce`, payload ?? {})

function compactMetricQuery(query?: RunMetricQuery) {
  if (!query) return undefined
  const params: Record<string, string | number> = {}
  if (query.names?.length) {
    params.names = query.names.join(',')
  }
  if (query.startStep !== undefined) {
    params.startStep = query.startStep
  }
  if (query.endStep !== undefined) {
    params.endStep = query.endStep
  }
  if (query.limit !== undefined) {
    params.limit = query.limit
  }
  if (query.downsample !== undefined) {
    params.downsample = query.downsample
  }
  return params
}

File: backend/dao/model/experiment.go
package model

import (
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"
)

type ExperimentVisibility string

const (
	ExperimentVisibilityPrivate ExperimentVisibility = "private"
	ExperimentVisibilityAccount ExperimentVisibility = "account"
)

type ExperimentRunStatus string

const (
	ExperimentRunStatusPending    ExperimentRunStatus = "pending"
	ExperimentRunStatusRunning    ExperimentRunStatus = "running"
	ExperimentRunStatusSucceeded  ExperimentRunStatus = "succeeded"
	ExperimentRunStatusFailed     ExperimentRunStatus = "failed"
	ExperimentRunStatusTerminated ExperimentRunStatus = "terminated"
	//nolint:misspell // Keep the existing API value spelling for compatibility.
	ExperimentRunStatusCancelled ExperimentRunStatus = "cancelled"
)

type Experiment struct {
	gorm.Model
	Name        string               `json:"name" gorm:"type:varchar(256);not null;index:idx_experiments_scope_name"`
	Description string               `json:"description" gorm:"type:text"`
	UserID      uint                 `json:"userID" gorm:"not null;index:idx_experiments_scope_name"`
	User        User                 `json:"-" gorm:"foreignKey:UserID"`
	AccountID   uint                 `json:"accountID" gorm:"not null;index:idx_experiments_scope_name"`
	Account     Account              `json:"-" gorm:"foreignKey:AccountID"`
	Visibility  ExperimentVisibility `json:"visibility" gorm:"type:varchar(32);not null;default:private"`
	Tags        datatypes.JSONMap    `json:"tags" gorm:"type:jsonb;comment:实验标签"`
}

type ExperimentRun struct {
	gorm.Model
	ExperimentID       uint                `json:"experimentID" gorm:"not null;index"`
	Experiment         Experiment          `json:"-" gorm:"foreignKey:ExperimentID"`
	ParentRunID        *uint               `json:"parentRunID" gorm:"index;comment:来源实验 Run ID"`
	SourceCheckpointID *uint               `json:"sourceCheckpointID" gorm:"index;comment:来源 checkpoint ID"`
	JobID              *uint               `json:"jobID" gorm:"index"`
	JobName            string              `json:"jobName" gorm:"type:varchar(256);index;comment:绑定的作业名"`
	RunName            string              `json:"runName" gorm:"type:varchar(256);not null"`
	Status             ExperimentRunStatus `json:"status" gorm:"type:varchar(32);not null;default:pending;index"`
	UserID             uint                `json:"userID" gorm:"not null;index"`
	AccountID          uint                `json:"accountID" gorm:"not null;index"`
	RunTokenHash       string              `json:"-" gorm:"type:varchar(128);not null"`
	Hyperparams        datatypes.JSONMap   `json:"hyperparams" gorm:"type:jsonb;comment:超参数"`
	CodeSnapshot       datatypes.JSONMap   `json:"codeSnapshot" gorm:"type:jsonb;comment:代码快照"`
	DataSnapshot       datatypes.JSONMap   `json:"dataSnapshot" gorm:"type:jsonb;comment:数据快照"`
	ImageSnapshot      datatypes.JSONMap   `json:"imageSnapshot" gorm:"type:jsonb;comment:镜像快照"`
	ResourceSnapshot   datatypes.JSONMap   `json:"resourceSnapshot" gorm:"type:jsonb;comment:资源快照"`
	CheckpointSnapshot datatypes.JSONMap   `json:"checkpointSnapshot" gorm:"type:jsonb;comment:checkpoint快照"`
	ReproduceSnapshot  datatypes.JSONMap   `json:"reproduceSnapshot" gorm:"type:jsonb;comment:复现快照"`
	Tags               datatypes.JSONMap   `json:"tags" gorm:"type:jsonb;comment:Run标签"`
	StartedAt          *time.Time          `json:"startedAt"`
	FinishedAt         *time.Time          `json:"finishedAt"`
}

type RunMetric struct {
	gorm.Model
	RunID          uint              `json:"runID" gorm:"not null;index:idx_run_metric_name_step,priority:1;uniqueIndex:idx_run_metric_client_record,priority:1"` //nolint:lll // GORM index tags are intentionally descriptive.
	Run            ExperimentRun     `json:"-" gorm:"foreignKey:RunID"`
	ClientRecordID *string           `json:"clientRecordID" gorm:"type:varchar(64);uniqueIndex:idx_run_metric_client_record,priority:2;comment:SDK离线记录ID"` //nolint:lll // GORM index tags are intentionally descriptive.
	Name           string            `json:"name" gorm:"type:varchar(256);not null;index:idx_run_metric_name_step,priority:2"`
	Step           int64             `json:"step" gorm:"not null;index:idx_run_metric_name_step,priority:3"`
	Value          float64           `json:"value" gorm:"not null"`
	Timestamp      time.Time         `json:"timestamp" gorm:"not null;index"`
	Context        datatypes.JSONMap `json:"context" gorm:"type:jsonb"`
}

type RunArtifact struct {
	gorm.Model
	RunID          uint              `json:"runID" gorm:"not null;index;uniqueIndex:idx_run_artifact_client_record,priority:1"`
	Run            ExperimentRun     `json:"-" gorm:"foreignKey:RunID"`
	ClientRecordID *string           `json:"clientRecordID" gorm:"type:varchar(64);uniqueIndex:idx_run_artifact_client_record,priority:2;comment:SDK离线记录ID"` //nolint:lll // GORM index tags are intentionally descriptive.
	Name           string            `json:"name" gorm:"type:varchar(256);not null"`
	Type           string            `json:"type" gorm:"type:varchar(64);not null;default:file"`
	Path           string            `json:"path" gorm:"type:text;not null"`
	SizeBytes      int64             `json:"sizeBytes" gorm:"not null;default:0"`
	SourceType     string            `json:"sourceType" gorm:"type:varchar(64);index;uniqueIndex:idx_run_artifact_source,priority:1;comment:来源类型"` //nolint:lll // GORM index tags are intentionally descriptive.
	SourceID       *uint             `json:"sourceID" gorm:"index;uniqueIndex:idx_run_artifact_source,priority:2;comment:来源记录ID"`
	Metadata       datatypes.JSONMap `json:"metadata" gorm:"type:jsonb"`
}
